### PR TITLE
July 2025 UPDATE: Improved patient tagging and OCR

### PR DESCRIPTION
**Pull request by GetWellClinic**
- improved prompts for category classification
- added workflow for extracting OCR
- improved cross-checking patient names
- improved accuracy of OCR with change in model
- improved accuracy of patient identification

## Summary by Sourcery

Improve OCR text extraction and patient tagging accuracy, and enhance installation scripts with maintenance automation.

New Features:
- Add PDF form field extraction to OCR processor to capture embedded form data.
- Parse HTML search results into structured JSON in patient filtering workflow.
- Include weekly system reboot cron jobs and automate Docker/group user additions in install scripts.

Enhancements:
- Refine regex logic for patient name, date of birth, and HIN parsing and cross-checking.
- Normalize and format LLM prompts with consistent spacing.
- Add fallback splitting logic for HIN matching and simplify HTML response handling.
- Extend date conversion to support multiple date formats and output dual formats.

Documentation:
- Update README version and hardware requirements, clarify default unidentified patient tagging ID field.
- Add troubleshooting sections for AI-MOA service reboot on kernel upgrades and PDF skipping behavior.

Chores:
- Bump version numbers across install and Docker scripts.
- Suppress permission errors in service startup and fix scripts.
- Expose new PDF extraction function in utils init and EMR workflow mappings.

### CODE REVIEW

File: src/processors/utils/ocr.py [hunk 1]
```diff
@@ -99,14 +101,39 @@ def extract_text_from_pdf_file(self):
     text = ''
     try:
         # Load the PDF from bytes
+        form_data = {}
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/utils/ocr.py:104]: @sourcery-ai please ignore this.

File: src/processors/patient_tagger/patient.py [hunk 2]
```diff
@@ -438,10 +457,12 @@ def get_patient_Html(self,type_of_query,query):
...
+    table = soup.find_all(class_="odd") + soup.find_all(class_="even")
 
-    return table
+    if table:
+        return response.text
+    else:
+        return
 
 
 def filter_results(self):
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:472]: @sourcery-ai please ignore this.

File: src/processors/patient_tagger/patient.py [hunk 3]
```diff
@@ -438,10 +457,12 @@ def get_patient_Html(self,type_of_query,query):
 
     soup = BeautifulSoup(response.text, 'html.parser')
 
-    table = soup.find_all(class_="odd")
-    table += soup.find_all(class_="even")
+    table = soup.find_all(class_="odd") + soup.find_all(class_="even")
 
-    return table
+    if table:
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:466]: @sourcery-ai please ignore this.

File: src/processors/patient_tagger/patient.py [hunk 4]
```diff
@@ -212,10 +190,21 @@ def get_patient_hin(self):
     # Remove leading and trailing letters
     query = re.sub(r'^[A-Za-z]+|[A-Za-z]+$', '', query)
 
-    search_match = re.search(re.escape(query), self.ocr_text)
-    
-    if search_match:
-        return self.get_patient_Html_Common(self,query,type_of_query)
+    result, data = self.get_patient_Html_Common(self,query,type_of_query)
+
+    if result:
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:195]: @sourcery-ai please ignore this.

File: src/processors/patient_tagger/patient.py [hunk 5]
```diff
@@ -746,19 +854,31 @@ def compare_name_with_text(self, data, text):
...
             name_parts = name.split()          # Split the name into parts
             flag = False
-            
 
             # Check if all part length
             if all(len(part) <= 3 for part in name_parts):
-                # We check if every part in name_parts has a match in OCR
-                if all(re.search(r'\b' + re.escape(part) + r'\b', text, re.IGNORECASE) for part in name_parts):
+                # Check if every part except for one in the name_parts has a match in OCR
+                missing_parts_count = sum(1 for part in name_parts if not re.search(r'\b' + re.escape(part) + r'\b', text, re.IGNORECASE))
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:865]: @sourcery-ai please ignore this.

File: src/processors/patient_tagger/patient.py [hunk 6]
```diff
@@ -438,10 +457,12 @@ def get_patient_Html(self,type_of_query,query):
 
     soup = BeautifulSoup(response.text, 'html.parser')
 
-    table = soup.find_all(class_="odd")
-    table += soup.find_all(class_="even")
+    table = soup.find_all(class_="odd") + soup.find_all(class_="even")
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:464]: @sourcery-ai please ignore this.

File: src/processors/patient_tagger/patient.py [hunk 7]
```diff
@@ -270,26 +259,56 @@ def convert_date(self,query):
...
+    if match_dd_mm:
+        query = match_dd_mm.group()
+        query = re.sub(r'[-/\s]', ' ', query)
+        day, month, year = query.split()
+    elif match_yyyy:
+        query = match_yyyy.group()
+        query = re.sub(r'[-/\s]', ' ', query)
+        year, month, day = query.split()
+    elif matchText:
+        query = matchText.group()
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:302]: @sourcery-ai This was added to handle an incorrect date-month format, and it will be addressed at the EMR end. Please ignore this.

File: src/processors/patient_tagger/patient.py [hunk 8]
```diff
@@ -212,10 +190,21 @@ def get_patient_hin(self):
...
+    result, data = self.get_patient_Html_Common(self,query,type_of_query)
+
+    if result:
+        return True, data
+    else:
+        n_str = str(query)
+        mid_index = len(n_str) // 2
+        if mid_index >= 3:
+            part1 = f"{int(n_str[:mid_index])}"
+            part2 = f"%{int(n_str[mid_index:])}"
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:202]: @sourcery-ai Thank you for pointing this, its updated.

File: src/processors/patient_tagger/patient.py [hunk 9]
```diff
@@ -123,46 +125,16 @@ def get_patient_dob(self):
...
-
-    patternText = r'\b(?:January|February|March|April|May|June|July|August|September|October|November|December) \d{1,2}, \d{4}\b'
-    matchText = re.search(patternText, query)
-
-    if match:
-        query = match.group()
-    elif matchText:
-        query = matchText.group()
-        query = self.convert_date(self,query)
+    formatted_dates = self.convert_date(self,query)
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:134]: @sourcery-ai It will be always a non-empty list or a boolean value.

File: src/processors/patient_tagger/patient.py [hunk 10]
```diff
@@ -72,6 +72,8 @@ def get_patient_name(self):
 
     if query:
 
+        query = re.sub(r'([A-Za-z]),([A-Za-z])', r'\1, \2', query)
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py]: @sourcery-ai This has been updated to include multiple characters. Thank you for pointing this out.

File: src/processors/patient_tagger/patient.py [hunk 11]
```diff
@@ -270,26 +259,56 @@ def convert_date(self,query):
...
+        query = matchText.group()
+        query = re.sub(r'[-/\s]', ' ', query)
+        query = re.sub(r'[,]', '', query)
+        month, day, year = query.split()
+        month = months.get(month.capitalize(), None)
+    elif matchText_YYYY:
+        query = matchText_YYYY.group()
+        query = re.sub(r'[-/\s]', ' ', query)
+        year, month, day = query.split()
+        month = months.get(month.capitalize(), None)
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:311]: @sourcery-ai please ignore this.

File: src/processors/patient_tagger/patient.py [hunk 12]
```diff
@@ -270,26 +259,56 @@ def convert_date(self,query):
...
+    elif match_yyyy:
+        query = match_yyyy.group()
+        query = re.sub(r'[-/\s]', ' ', query)
+        year, month, day = query.split()
+    elif matchText:
+        query = matchText.group()
+        query = re.sub(r'[-/\s]', ' ', query)
+        query = re.sub(r'[,]', '', query)
+        month, day, year = query.split()
+        month = months.get(month.capitalize(), None)
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:306]: @sourcery-ai please ignore this.

File: src/processors/utils/ocr.py [hunk 13]
```diff
@@ -25,6 +25,8 @@
 from doctr.models import ocr_predictor
 import torch
 import requests
+import fitz
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/utils/ocr.py:28]: @sourcery-ai please ignore this.

File: install/Readme.md [hunk 14]
```diff
@@ -602,6 +603,33 @@ sudo mkdir /usr/share/oscar-emr/OscarDocument/oscar/incomingdocs/1/Refile
 sudo chown -R tomcat9:tomcat9 /usr/share/oscar-emr/OscarDocument/oscar/incomingdocs/1
 ```
 
+### AI-MOA periodically stops working completely ###
+
+This can happen when the Linux Ubuntu system does an automatic upgrade of the Linux kernel or the NVIDA video drivers with unattended upgrade packages. You can check if you can still access the ```nvidia-smi``` display and check the GPU memory usage.
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply install/Readme.md]: @sourcery-ai this has been updated, thank you.

File: install/install-aimoa.sh [hunk 15]
```diff
@@ -120,10 +120,12 @@ pip install -r "$AIMOA/src/requirements.txt"
 
 # Create Linux user and group for 'aimoa':
 /usr/sbin/useradd -m aimoa	# Requires home directory for google-chrome files
+/bin/newgrp aimoa
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply install/install-aimoa.sh:123]: @sourcery-ai please ignore this.

File: src/processors/utils/ocr.py [hunk 16]
```diff
@@ -99,14 +101,39 @@ def extract_text_from_pdf_file(self):
...
+                                if field_value:
+                                    field_value = field_value.strip('()')  # Remove parentheses from value
+                                else:
+                                    field_value = ''  # Empty fields should still be captured
+
+                                # Store the field name and value in the dictionary
+                                form_data[field_name] = field_value
+        
+        form_data_str = "\n".join([f"{field_name}: {field_value}" for field_name, field_value in form_data.items()])
+        self.ocr_text = form_data_str
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/utils/ocr.py:135]: @sourcery-ai will consider this suggestion later on, please ignore this for now.

File: src/processors/patient_tagger/patient.py [hunk 17]
```diff
@@ -746,19 +854,31 @@ def compare_name_with_text(self, data, text):
...
                             flag = True
+                    if len(part) == 3:
+                        match = re.search(r'\b' + re.escape(part) + r'\b', text, re.IGNORECASE)
+                        if match:
+                            parts_count += 1
+                            if parts_count >= 2:
+                                year = data.get('formattedDob').split('-')[0]
+                                match = re.search(r'\b[-/.]*' + re.escape(year) + r'[-/.]*\b', text, re.IGNORECASE)
+                                if match:
+                                    flag = True
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:885]: @sourcery-ai its already safe-guarded in the verification step, please ignore this.

File: src/processors/patient_tagger/patient.py [hunk 18]
```diff
@@ -478,15 +551,47 @@ def filter_results(self):
...
+
+                match = re.search(r'```json\n(.*?)```', text, re.DOTALL)
+
+                if match:
+                    text = match.group(1)
+
+            else:
+                text = matched_data_array[0]
+
+        json_pattern = r'\{[^{}]*\}'
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:590]: @sourcery-ai not expecting a nested json, please ignore this.

File: src/processors/patient_tagger/patient.py [hunk 19]
```diff
@@ -462,9 +483,61 @@ def filter_results(self):
...
-    table = self.config.get_shared_state('type_of_query_table')
-    if type_of_query is not None:
-        prompt = f"\n{table}.\n{self.ocr_text}.\n" + self.ai_prompts.get('get_patient_result_filter', '')
+    str_table = self.config.get_shared_state('type_of_query_table')
+
+    soup = BeautifulSoup(str_table, 'html.parser')
+
+    table = soup.find_all(class_="odd") + soup.find_all(class_="even")
+
+    result_table_json = []
```
[the remark this answers is not in the document]
Comment (justinjosephmkj) [reply src/processors/patient_tagger/patient.py:496]: @sourcery-ai it will be validated in the previous step, please ignore this.
